src/common/JSHandle.ts

```diff
@@ -134,11 +134,15 @@ export class JSHandle<T = unknown> {
 
   async evaluate<
     Params extends unknown[],
-    Func extends EvaluateFunc<[T, ...Params]> = EvaluateFunc<[T, ...Params]>
+    Func extends EvaluateFunc<[this, ...Params]> = EvaluateFunc<
+      [this, ...Params]
+    >
   >(
     pageFunction: Func | string,
-    ...args: EvaluateParams<Params>
-  ): Promise<Awaited<ReturnType<Func>>> {
+    ...args: Params
+  ): // @ts-expect-error Circularity here is okay because we only need the return
```

Comment:
having test types would be nice at this point? does TS generate proper d.ts if there are some errors?

Reply:
Yes, it still generates the types correctly. I assume this would break if the type is truly circular (meaning it returns `this` somehow). Regarding testing this, the test files that changed at the tests in some sense. The typing on the files that changed broke as soon as I implemented this, which makes sense because they were wrong before.